coderd/workspaceupdates.go

```diff
@@ -0,0 +1,313 @@
+package coderd
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/pubsub"
+	"github.com/coder/coder/v2/coderd/rbac"
+	"github.com/coder/coder/v2/coderd/rbac/policy"
+	"github.com/coder/coder/v2/coderd/util/slice"
+	"github.com/coder/coder/v2/coderd/wspubsub"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+)
+
+type UpdatesQuerier interface {
+	GetAuthorizedWorkspacesAndAgentsByOwnerID(ctx context.Context, ownerID uuid.UUID, prep rbac.PreparedAuthorized) ([]database.GetWorkspacesAndAgentsByOwnerIDRow, error)
+	GetWorkspaceByAgentID(ctx context.Context, agentID uuid.UUID) (database.Workspace, error)
+}
+
+type workspacesByID = map[uuid.UUID]ownedWorkspace
+
+type ownedWorkspace struct {
+	WorkspaceName string
+	Status        proto.Workspace_Status
+	Agents        []database.AgentIDNamePair
+}
+
+// Equal does not compare agents
+func (w ownedWorkspace) Equal(other ownedWorkspace) bool {
+	return w.WorkspaceName == other.WorkspaceName &&
+		w.Status == other.Status
+}
+
+type sub struct {
+	ctx      context.Context
+	cancelFn context.CancelFunc
+
+	mu       sync.RWMutex
+	userID   uuid.UUID
+	ch       chan *proto.WorkspaceUpdate
+	prev     workspacesByID
+	readPrep rbac.PreparedAuthorized
+
+	db     UpdatesQuerier
+	ps     pubsub.Pubsub
+	logger slog.Logger
+
+	psCancelFn func()
+}
+
+func (s *sub) handleEvent(ctx context.Context, event wspubsub.WorkspaceEvent, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	switch event.Kind {
+	case wspubsub.WorkspaceEventKindStateChange:
+	case wspubsub.WorkspaceEventKindAgentConnectionUpdate:
+	case wspubsub.WorkspaceEventKindAgentTimeout:
+	case wspubsub.WorkspaceEventKindAgentLifecycleUpdate:
+	default:
+		if err == nil {
+			return
+		} else {
+			// Always attempt an update if the pubsub lost connection
+			s.logger.Warn(ctx, "failed to handle workspace event", slog.Error(err))
+		}
+	}
+
+	rows, err := s.db.GetAuthorizedWorkspacesAndAgentsByOwnerID(ctx, s.userID, s.readPrep)
+	if err != nil {
```

Comment:
can this thing return `sql.ErrNoRows`? If so you need to handle it.

Reply:
Nope, only queries that return a single row will return `ErrNoRows`.